test/e2e/test/registry.go

```diff
@@ -0,0 +1,152 @@
+var retriable = func(err error) bool {
+	// Do not retry on permission errors or if the resource is not found.
+	return !(client.IsForbidden(err) || client.IsUnauthorized(err) || client.IsNotFound(err))
+}
+
+// WithDigestOrDie returns a fully qualified image name with the version (tag) replaced by the image digest.
+// Retries are attempted on a best effort basis. This function panic once the maximum retries is reached.
+func WithDigestOrDie(image container.Image, tag string) (withDigest string) {
+	err := retry.OnError(
+		backoff,
+		retriable,
+		func() error {
+			digest, err := getDigest(image, tag)
+			if err != nil {
+				logf.Log.Error(
+					err,
+					"Error while fetching image digest",
+					"image", image,
+					"tag", tag,
+					"retriable", retriable(err),
+				)
+				return err
+			}
+			withDigest = fmt.Sprintf("%s/%s@%s", container.DefaultContainerRegistry, image, digest)
+			return nil
+		},
+	)
+	if err != nil {
+		// We eventually panic as it is likely to prevent any subsequent test to run.
+		panic(err)
+	}
+	return
+}
+
+func getDigest(image container.Image, tag string) (string, error) {
+	imagesDigestsSingleton.lock.Lock()
+	defer imagesDigestsSingleton.lock.Unlock()
+
+	taggedImage := fmt.Sprintf("%s:%s", image, tag)
+	imageDigest := imagesDigestsSingleton.images[taggedImage]
+	if imageDigest != "" {
+		logf.Log.Info("Reusing image digest from cache", "image", image, "tag", tag, "digest", imageDigest)
+		return imageDigest, nil
+	}
+
+	repo, err := name.NewRepository(fmt.Sprintf("%s/%s", container.DefaultContainerRegistry, image))
+	if err != nil {
+		return "", err
+	}
+
+	auth, err := authn.DefaultKeychain.Resolve(repo.Registry)
+	if err != nil {
+		return "", err
+	}
+
+	// Construct an http.Client that is authorized to pull images
+	scopes := []string{repo.Scope(transport.PullScope)}
+	t, err := transport.New(repo.Registry, auth, http.DefaultTransport, scopes)
+	if err != nil {
+		return "", err
+	}
+	client := &http.Client{Transport: t}
+
+	// URL to fetch the image manifest.
+	uri := &url.URL{
+		Scheme: repo.Registry.Scheme(),
+		Host:   repo.Registry.RegistryStr(),
+		Path:   fmt.Sprintf("/v2/%s/manifests/%s", image, tag),
+	}
+
+	req, err := http.NewRequest("GET", uri.String(), nil) //nolint:noctx
+	if err != nil {
+		return "", err
+	}
+
+	// Adding "application/vnd.docker.distribution.manifest.v2+json" is required to get the correct image Digest.
+	req.Header.Add("Accept", string(types.DockerManifestSchema2))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	// Assert that we get a 200, otherwise attempt to parse body as a structured error.
+	if err := transport.CheckError(resp, http.StatusOK); err != nil {
+		return "", err
+	}
+
+	// Digest is not in the response body, it's in the header "Docker-Content-Digest"
+	digest := resp.Header.Get("Docker-Content-Digest")
+	if digest == "" {
+		return "", errors.New("No image digest in response")
+	}
```

Comment:
Is there a particular reason why this is not simply: 

```go
	withRegistry := fmt.Sprintf("%s/%s", container.DefaultContainerRegistry, taggedImage)
	digest, err := crane.Digest(withRegistry)
	if err != nil {
		return "", err
	}
```

Reply:
I spent too much time figuring out how the protocol works behind the scenes, I forgot to check if this was implemented elsewhere in the library 😄